modules/api/php/endpoints/subprojects.class.inc

```diff
@@ -0,0 +1,149 @@
+     *
+     * @return boolean true if access is permitted
+     */
+    private function _hasAccess(\User $user) : bool
+    {
+        return !($user instanceof \LORIS\AnonymousUser);
+    }
+
+    /**
+     * Return which methods are supported by this endpoint.
+     *
+     * @return array supported HTTP methods
+     */
+    protected function allowedMethods() : array
+    {
+        return ['GET'];
+    }
+
+    /**
+     * Versions of the LORIS API which are supported by this
+     * endpoint.
+     *
+     * @return array a list of supported API versions.
+     */
+    protected function supportedVersions() : array
+    {
+        return [
+            'v0.0.4-dev',
+        ];
+    }
+
+    /**
+     * Handles a request starts with /projects/
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        $user = $request->getAttribute('user');
+        if (!$this->_hasAccess($user)) {
+            return new \LORIS\Http\Response\JSON\Unauthorized();
+        }
+
+        $pathparts = $request->getAttribute('pathparts');
+
+        if (count($pathparts) === 1) {
+            switch ($request->getMethod()) {
+            case 'GET':
+                return $this->_handleGET($request);
+
+            case 'OPTIONS':
+                return (new \LORIS\Http\Response())
+                    ->withHeader('Allow', $this->allowedMethods());
+            default:
+                return new \LORIS\Http\Response\JSON\MethodNotAllowed(
+                    $this->allowedMethods()
+                );
+            }
+        }
+
+        // Delegate to subproject specific endpoint.
+        $subproject_name =  urldecode($pathparts[1] ?? '');
+        if (!in_array(
+            $subproject_name,
+            array_values(\Utility::getSubprojectList())
+        )
+        ) {
+            return new \LORIS\Http\Response\JSON\NotFound();
+        }
+
+        $endpoint = new Subproject\Subproject($subproject_name);
+        // Removing `/subprojects/<subproject_name>` from pathparts.
+        $pathparts = array_slice($pathparts, 2);
+        $request   = $request->withAttribute('pathparts', $pathparts);
+
+        return $endpoint->process($request, $endpoint);
+    }
+
+    /**
+     * Returns a list of subprojects for this LORIS instance.
+     *
+     * @return ResponseInterface
+     */
+    private function _handleGET() : ResponseInterface
+    {
+        if (!isset($this->_cache)) {
+            // The subprojects list should be related to the user in the request
+            // attributes.
+            $DB           = \NDB_Factory::singleton()->database();
+            $subprojects  = $DB->pselect("SELECT * FROM subproject", []);
+            $this->_cache = (new \LORIS\api\Views\Subprojects($subprojects))
+                ->toArray();
```

Comment:
This should use a dataprovisioner instead of querying the database directly. The dataprovisionner should return a Traversable of Subproject objects that would be passe to the view class. 
In the view class, getting the the visitlabels fro each subproject should be something like `$subproject->getVisitLabels();`

Using the dataprovisioner will allow us to add  UserProjectMatch to filter out the subprojects not available to the user.
Creating a Subproject "model" class will create a single source of truth (to answer Dave's comment about using the Utility class to to get visits for a subproject)